### ISSUE TITLE
App Security findings in IntuneMAM SDK

### ISSUE THREAD

**Issue by Nookaraju** [msintuneappsdk/ms-intune-app-sdk-ios#38]
**Describe the bug:**
When our app(linked with IntuneMAM SDK) is scanned by a security tool for app security findings, there are some issues that were from Intune MAM SDK.

**To Reproduce**
List of issues:
**1. MD5 usage* 
Issue details: 

The following code locations used the deprecated the CC_MD functions to generate a message digest:

```
[CUTCategories md5] calls _CC_MD5()
[CMARCryptoUtils md5Hash:] calls _CC_MD5()
```

**2. Switch from NSCoding to NSSecureCoding** 

NSCoding protocol used in binary <>.app/Frameworks/IntuneMAM.framework/IntuneMAM

Description:
NSSecureCoding should be used.

(**_FIXED FROM ADAL 4.0_**) **3. Switch from Deprecated UIWebView API to WKWebView**  -- 

```
-[ADAuthenticationViewController loadView:] calls -[_OBJC_CLASS_$_UIWebView initWithFrame:]
```

 _(**ADAL Lib**)_ **4. Null Initialization Vector Used **  --

The following code locations used the CommonCrypto APIs to perform AES encryption with a NULL Initialization Vector:

`-[ADBrokerKeyHelper decryptBrokerResponse:key:size:error:] calls _CCCrypt(_, _, _, _, _, NULL, _)
`

**5. Broken Cryptography API (SHA1) used**

The following code locations used CC_SHA1 functions to generate a message digest:
```
-[ADPkeyAuthHelper computeThumbprint:isSha2:] calls _CC_SHA1() .      _(**ADAL Lib**)_
-[MSIDPkeyAuthHelper computeThumbprint:isSha2:] calls _CC_SHA1()
-[MSIDExtensions msidSHA1] calls _CC_SHA1()
```

**Expected behavior:**
IntuneMAM framework should not have app security issues

**Intune App SDK for iOS (please complete the following information):**
- What version of the Intune SDK are you using? Are you using the latest version?[
[DEV]: 11.0.2
- What platform is your app based in (Java, Xamarin based, Cordova, etc)?
[DEV]: Swift 
- Who is the customer?

**Additional context:**

**Comment by Nookaraju** [comment]
Any update on this?

**Comment by Nookaraju** [comment]
@Kyle-Reis These security issues are holding our app release, can you please take look and give us update?

**Comment by Kyle-Reis** [comment]
Hey @Nookaraju, we’re investigating items 1 and 2 on the list. The remaining items all appear to be coming from ADAL, so you should file an issue [here](https://github.com/AzureAD/azure-activedirectory-library-for-objc/issues). 

The MD5 hashes we generate are meant to obfuscate things like user’s UPN and URL endpoints in our logs, and aren’t used for encrypting actual corporate data. We’ll look into whether there is another lightweight obfuscation method available.

We only use the NSCoding protocol to serialize/deserialize to the iOS keychain, which is protected by the OS and only accessible to apps signed by the same publisher. We will look into moving to NSSecureCoding for an added layer of protection.

I don’t think you should consider either of the items above blockers to release, but if you feel differently please let us know :)

**Comment by Nookaraju** [comment]
@Kyle-Reis Thanks for your reply. I have created separate issue for ADAL([link](https://github.com/AzureAD/azure-activedirectory-library-for-objc/issues/1432))

Just wanted to know, when we can expect the fixes for first 2 issues from Intune MAM SDK.

Thank you.

**Comment by Nookaraju** [comment]
Added new item 6. App Uses Deprecated API to Authenticate Through a Web Service

**Comment by Kyle-Reis** [comment]
Hey @Nookaraju, the changes to address items 1 and 2 are now being reviewed. Item #6 looks like it is also in ADAL, and not in Intune.

**Comment by Nookaraju** [comment]
@Kyle-Reis Thanks for your quick reply. I have moved item 6 to ADAL lib.

**Comment by Nookaraju** [comment]
@Kyle-Reis Item 3 is very important now as Apple will stop accepting the app submission if we use UIWebView.

Check this out, got it from Apple. 
![11055304-a7ac-4c40-886f-5b1b0b2085c4](https://user-images.githubusercontent.com/2161122/64254549-450dd580-cf3d-11e9-83b9-7cfc8dffea3c.JPG)

**Comment by Kyle-Reis** [comment]
Hey @Nookaraju, item 3 is an ADAL 2 issue. However, the latest version of our SDK supports ADAL 4, so I would try upgrading Intune to the latest and upgrading to ADAL 4 as I believe they've moved from UIWebView to WKWebView in that release. Note: the current 11.1.2 release of our Swift framework does have a known issue with building for simulator (see #47). We've already addressed that issue and will make the fix available soon, but you should be able to build and test on physical devices without issue.

**Comment by Nookaraju** [comment]
@Kyle-Reis Upgrading to ADAL 4 fixed the UIWebView issue.

But Security Tool found one more issue:

The following code locations within the App use Security.framework functions to generate cryptographic key pairs:
-[CMAREncryptionProvider generateKeysWithPublicKeyTag:PrivateKeyTag:AccessGroup:] calls _SecKeyGeneratePair()

**Description:**
The App uses Apple's Security framework to generate cryptographic key pairs, but does not leverage the frameworks' Secure Enclave features introduced with iOS 13, to securely store private keys.

Starting with iOS 13, developers can add an extra layer of security by leveraging the hardware based key manager located in the Secure Enclave, available in any iOS devices with an A7 chip or newer.

When a private key is stored in the Secure Enclave, the key never actually leaves the Secure Enclave and your App or the OS itself never has to store a plain text version of the key in memory, in order to use it. Instead, your App instructs the Secure Enclave to create the key, securely store it, and perform cryptographic operations with it, and then only receives the output of these operations (such as encrypted data, or the result of a signature verification).

**Recommendation**
Leverage the Secure Enclave capabilities when performing any cryptographic operation on the device, in order to add an extra layer of security for the App's private keys.

Please refer to Apple's official documentation at https://developer.apple.com/documentation/security/certificate_key_and_trust_services/keys/storing_keys_in_the_secure_enclave for more details.

**Comment by Kyle-Reis** [comment]
Hello @Nookaraju, Items 1 and 2 have been addressed in our [latest release](https://github.com/msintuneappsdk/ms-intune-app-sdk-ios/releases/latest)(11.2.0). Note: the tool will still pick up on MD5 function calls as we needed to preserve backwards compat. Once adoption of newer SDKs has reached critical mass, we will drop support for MD5.

As for the Secure Enclave, we did some investigation into using it some time ago, and found that it would not suit our needs, as encrypted data needs to be shared between applications of different publishers, which the Secure Enclave does not currently support.
